### ISSUE TITLE
Native Sidecars draining erroneously sending connection: close for outbound requests

### ISSUE THREAD

**Issue by Stono** [istio/istio#51855]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

When using native sidecars, the following lifecycle is added:

```
    lifecycle:
      preStop:
        exec:
          command:
          - pilot-agent
          - request
          - --debug-port=15020
          - POST
          - drain
```

The `preStop` event fires when the `Pod` enters a terminating state.  At this point, the proxy starts sending `connection: close` for **outbound** requests, eg the app will still be making outbound calls at this point, but getting a `connection: close` back.  

You can see this by doing the following:

- Add a `preStop: sleep 10` to your `app` container
- Exec into the pod, and setup a `watch -n 1 curl <whatever>`
- `kubectl delete the pod`
- Immediately see the `connection: close` headers being returned

This causes some of our http pool implementations to quite literally implode! :D 

So far, I've only got the issue to manifest when the remote destination does not have a sidecar too, and the app has long lived http2 connections.  For example:

`app1 | app1-proxy ---------> remote app (no sidecar)`

I do not believe we should be seeing `connection: close` for these outbound requests.

### Version

```Text
1.22.2
```

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [X] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Comment by howardjohn** [comment]
This seems to be a bug in Envoy.

`curl -X POST localhost:15000/'drain_listeners?inboundonly'`: works

`curl -X POST localhost:15000/'drain_listeners?inboundonly&graceful'`: does not (that is, it applies to outbound traffic)

**Comment by Stono** [comment]
As a mitigation, i've removed the shutdown hook from our injection template.
We already have some reasonable guards to aid with graceful shutdowns that have historically worked _ok_ for us.

**Comment by howardjohn** [comment]
The logic in envoy:

```
  if (graceful) {
    // Ignore calls to /drain_listeners?graceful if the drain sequence has
    // already started.
    if (!server_.drainManager().draining()) {
      server_.drainManager().startDrainSequence([this, stop_listeners_type, skip_exit]() {
        if (!skip_exit) {
          server_.listenerManager().stopListeners(stop_listeners_type, {});
        }
      });
    }
  } else {
    server_.listenerManager().stopListeners(stop_listeners_type, {});
  }
```

You can see in graceful+skip_exit+inboundonly, the `stop_listeners_type=inboundonly` is not used anywhere

**Comment by howardjohn** [comment]
Opened https://github.com/envoyproxy/envoy/issues/35020

**Comment by ramaraochavali** [comment]
Is n't this a problem for regular sidecars as well?

**Comment by Stono** [comment]
No, the prestop hook to drain is only added when native sidecars is enabled

**Comment by ramaraochavali** [comment]
But the same logic gets executed when proxy gets SIGTERM

**Comment by Stono** [comment]
If that is true then yes, it is a problem there too

**Comment by dwj300** [comment]
> Is n't this a problem for regular sidecars as well?

@ramaraochavali I think it is, right? Let's say `terminationDrainDuration` is 5min, and `drainDuration` is 45s. Presumably, a workload could send a new outbound request after one minute, and the outbound listener would already be shutdown...

**Comment by ramaraochavali** [comment]
yes

**Comment by tbrasser** [comment]
> If that is true then yes, it is a problem there too. The reason I might not have seen it historically is for years we've had a `pilot-agent` wrapper script that blocked SIGTERM getting to the proxy and did a `while` loop looking for `netstat` listening ports from the main app. We removed that "hacky" script as part of the move to native sidecar containers.

Can we get that "hacky" script back until this is fixed in envoy?

**Comment by Stono** [comment]
That hacky script was never part of istio.  It was my hacky script I baked into my custom proxy image.

**Comment by Stono** [comment]
Istio maintainers - can we remove the stale label?  This issue is absolutely an issue! :)

**Comment by chlunde** [comment]
We added a sleep preStop hook (https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3960-pod-lifecycle-sleep-action/README.md) to our pods to ensure the app container does not stop before the istio sidecar stops. This fixed one set of issues getting us closer to 0% failure rate in our mesh, but I believe some of our teams are hitting this issue now instead, most likely for apps that perform outbound connection that are initiated by the app "on its own", not to handle an inbound request.

**Comment by grzesuav** [comment]
hi, till it is resolved, what would be the "right" preStop hook to be added ? I am interested in using this feature but this seems like a blocker, I wonder if I can somehowe replace the hook to use `curl -X POST localhost:15000/'drain_listeners?inboundonly' equivalent

**Comment by AndresPinerosZen** [comment]
Could someone please explain why this is happening? Aren't native sidecars supposed to start their termination process only after the main containers have terminated? I'm just going to start experimenting with Native Sidecars and I thought that's how they were supposed to work

**Comment by hzxuzhonghu** [comment]
The difference lies on `skip_exit`, from the ctx above, it seems envoy doesnot work as expected https://github.com/istio/istio/issues/51855#issuecomment-2203905301

So now the right way is to continue using original drain way rather than native sidecar

**Comment by Stono** [comment]
Hmm not sure? 
I use native sidecars fine now, I just simply removed the `preStop` hook that istio adds when native sidecars is enabled which is what triggers this behaviour on pod stop.

**Comment by michaelbannister** [comment]
@AndresPinerosZen it seems that the preStop hook is run when the pod begins terminating. I found it mentioned in [this Layer5 blog about 1.29 changes](https://layer5.io/blog/kubernetes/kubernetes-129-highlights-features-and-deprecations). It's not mentioned on the main [Kubernetes docs page about Sidecar containers](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) and I couldn't find it elsewhere in Kubernetes docs proper, only [in the Sidecar Containers KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/README.md#termination-of-containers).

**Comment by PatTheSilent** [comment]
@Stono How did you remove the preStop hook? Just manually removed that part from the ConfigMap or is there some smarter way that escapes me?

**Comment by Stono** [comment]
Aye, it's crude, i just removed it from the template.

**Comment by chlunde** [comment]
> Immediately see the connection: close headers being returned
> This causes some of our http pool implementations to quite literally implode! :D

Is the problem _only_ that outbound connection _pooling_ stops working? Unlike JDBC pooling, HTTP connection pooling typically do not proactively open new connections, in my experience? Which library is this @Stono ?

What's the ideal timeline here?

1. Stop initiated
2. Sidecar _can_ start draining inbound keep-alive connections. Sidecar and app should still accept and process inbound requests. Apparently, don't do connection close for outbound requests.
3. Sidecar and app pod wait until the mesh is configured not to send any traffic to this service. There's no real event signalling this.
4. App pod can now get term signal (this could in practice be a after 5-10 second delay to cover 99.9xx% of use cases). This might initiate new outbound request so the istio sidecar should still process them
5. App pod terminates
6. Sidecar can now terminate

**Comment by rtmie** [comment]
Why is the resolution for this being punted to Envoy?

According to [sidecar KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/README.md#termination-of-containers)

1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
   the Pod's termination grace period countdown.
2. Once the last primary container terminates, the last started sidecar container is notified by
   sending a `SIGTERM` signal.
3. The next sidecar (in reverse order) is notified by sending a `SIGTERM` signal after the previous
   sidecar container terminates.
4. This continues until all sidecar containers have terminated, or the Pod's termination grace period
   expires.
5. In the latter case, all remaining containers are notified by a `SIGTERM`, followed by a fixed
   grace period of 2s and finally terminated.
6. The Pod will be terminated after that.


On this basis would it not make for since for pilot-agent to manage - ie no-op the preStop call and only make the drain call when it gets the SIGTERM?
In effect we can in fact shutdown immediately as all the primary containers are already terminated at that time and as long as the proxy is at the start of the initContainer list it will only get a SIGTERM after all subsequent initContainers have already received it.

**Comment by howardjohn** [comment]
Just to get on the same page on terminology - drain in this context means "encourage clients to go away gracefully". i.e. send `connection: close`. 

When it gets a SIGTERM, the application is gone. It doesn't make sense to drain now -- we cannot handle ANY requests, so we don't want to just encourage clients go away - we want to stop accepting any requests.

A correct but not ideal implementation would be to only handle SIGTERM and shutdown immediately. However, this doesn't give clients a period where they can gracefully disconnect.

The (IMO) optimal solution is:
* When pod shutdown start draining - encouraging clients to find other noty-shutting-down instances
* When the application terminates, exit immediately

**Comment by howardjohn** [comment]
BTW not sure if the concern was behavior or timing but this was merged in Envoy: https://github.com/envoyproxy/envoy/pull/37873

**Comment by keithmattix** [comment]
> BTW not sure if the concern was behavior or timing but this was merged in Envoy: [envoyproxy/envoy#37873](https://github.com/envoyproxy/envoy/pull/37873)

Unfortunately, it was reverted due to some flaky tests on main 😢 I'm working to repro locally so we can get it back in ASAP

**Comment by rtmie** [comment]
@howardjohn thanks for response and apologies for loose terminology!

So for my understanding of the desired behaviour (pending the requested changes in envoy) - is the below correct?

1. Pod gets initial "shutdown signal" - preStop hooks are triggered - application container and native-sidecar proxy
2. application starts doing whatever stuff it does in its prestop and pilot-agent puts envoy into draining (inbound-only&graceful-shutdown) 
3. Proxy no longer accepts new inbound requests, but maintains existing inbound connections until they close (pending the completion of preStop in app container or pod terminationGracePeriod)
4. During this period - new outbound requests can be made from the app container if triggered by it's pre-stop logic?
5. App container(s) preStop(s) complete - native sidecars start to get SIGTERM from kubelet in reverse order (progresses based on termination of each native sidecar) up to and including istio-proxy - at which point any remaining connections are gone.


WRT "A correct but not ideal implementation would be to only handle SIGTERM and shutdown immediately. However, this doesn't give clients a period where they can gracefully disconnect."
Understood - I guess my query was with this model - will the combination of application preStop processing and pod readiness handling give some reasonable level of protection to clients? i.e. to what extent is it istio's responsibility to ensure that app container gives a reasonable shutdown behaviour to its clients. However I do agree with ideal behaviour as you outlined. 

Wonder if these behaviour choices should be configurable in the sidecar injector based in a value in the  mesh profile?

**Comment by howardjohn** [comment]
The gap here is (3). The proxy does not stop accepting inbound requests - it just tells clients to not re-use the connection for further requests (so it will, hopefully, connect somewhere else).


> to what extent is it istio's responsibility to ensure that app container gives a reasonable shutdown behaviour to its clients

IMO this is up to the application to ensure they gracefully shutdown and do not just immediately quit. If their app cannot do it, its a pretty easy ~3 lines of config to put a `sleep` prestop

**Comment by rtmie** [comment]
"The gap here is (3). The proxy does not stop accepting inbound requests - it just tells clients to not re-use the connection for further requests (so it will, hopefully, connect somewhere else)."


ah gotcha - that's a subtle but important point I missed!!! And thats what we need the requested change in Envoy to support.

**Comment by rtmie** [comment]
wanted to refer back to this point

"these behaviour choices should be configurable in the sidecar injector based in a value in the mesh profile?"

i.e. the mesh operator should be able to configure the behaviour here between a "hard" or "soft" behaviour.

the istio proxy will be the first native sidecar and hence based on the KEP it will not get sigterm until application containers and all other native sidecars have terminated. So in many cases the preStop injected into istio native sidecar will not be necessary.

Making this behaviour of the sidecar injector configurable puts the decision on this behaviour into the hands of the mesh operator.

**Comment by howardjohn** [comment]
Why do you _not_ want to send `connection: close` headers to clients during shutdown to tell them to backoff?

**Comment by keithmattix** [comment]
Resubmitted #38430 and that merged, so this should be fixed in Istio 1.26. @Stono Any chance you want to give this a try using a [dev build](https://github.com/istio/istio/wiki/Dev%20Builds)? The latest master commit should have it since istio/proxy#6160 merged

**Comment by schahal** [comment]
> so this should be fixed in Istio 1.26

Is it confirmed this is in the v1.26 release (I didn't see a mention of it in the Release Notes, and wonder how I can confirm aside from upgrading and testing).

**Comment by keithmattix** [comment]
Yes this is in 1.26; I just forgot to call it out in release notes since it was an envoy only feature that didn't really require istio changes. If you have workloads that you know tend to trigger it, that'll be the best way to test. I wasn't able to get to it outside of the envoy integratio. Tests

**Comment by keithmattix** [comment]
Are folks seeing this behavior fixed? Let me know if so and I can close this issue

**Comment by kfaseela** [comment]
> Are folks seeing this behavior fixed? Let me know if so and I can close this issue

@Stono : Do you have any updates after testing this fix on 1.26? We are also hitting this internally on 1.25 and would like to just get some heads up whether 1.26 really has fixed the issue.

**Comment by schahal** [comment]
Missed these last few messages. I specifically upgraded to v1.26 in mid-May to pull in this exact fix:

It worked! Thank you!

TL;DR of why this was important for me: with native sidecars, the "preStop" command for sidecars gets run **_immediately_** when pod begins terminating (and the SIGTERM to the sidecars only get sent when the main application container is exited)... So with the "drain" bug the main app was failing some of the tasks during termination (a rare case that gets hit in my app, which is a job-type app)

**Comment by kfaseela** [comment]
> Missed these last few messages. I specifically upgraded to v1.26 in mid-May to pull in this exact fix:
> 
> It worked! Thank you!
> 
> TL;DR of why this was important for me: with native sidecars, the "preStop" command for sidecars gets run **_immediately_** when pod begins terminating (and the SIGTERM to the sidecars only get sent when the main application container is exited)... So with the "drain" bug the main app was failing some of the tasks during termination (a rare case that gets hit in my app, which is a job-type app)
> 

Thank you so much for confirming!
